Improper characters being used to construct URL in JUnitXmlTestRunResultProcessor

This was logged by continuous integration:

```
WARNING: Invocation of TestRunResultProcessor org.igniterealtime.smack.inttest.util.JUnitXmlTestRunResultProcessor@6562cc23 failed.
java.lang.IllegalArgumentException: Illegal character in fragment at index 45: https://xmpp.org/extensions/xep-0045.html#7.1 & 7.2.2
	at java.base/java.net.URI.create(URI.java:883)
	at org.igniterealtime.smack.inttest.util.JUnitXmlTestRunResultProcessor.process(JUnitXmlTestRunResultProcessor.java:215)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework.main(SmackIntegrationTestFramework.java:119)
Caused by: java.net.URISyntaxException: Illegal character in fragment at index 45: https://xmpp.org/extensions/xep-0045.html#7.1 & 7.2.2
	at java.base/java.net.URI$Parser.fail(URI.java:2913)
	at java.base/java.net.URI$Parser.checkChars(URI.java:3084)
	at java.base/java.net.URI$Parser.parse(URI.java:3128)
	at java.base/java.net.URI.<init>(URI.java:600)
	at java.base/java.net.URI.create(URI.java:881)
	... 2 more
```
It appears that some content that's used to generate a URL is not properly escaped.

--- a/src/main/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessor.java
+++ b/src/main/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessor.java
@@ -347,7 +347,7 @@ public class JUnitXmlTestRunResultProcessor implements SmackIntegrationTestFrame
         if (specification == null || specification.isBlank()) {
             return "";
         }
-        return specification.replaceAll("\\s", "").toUpperCase();
+        return specification.replaceAll("[\\s-]", "").toUpperCase();
     }
 
     public static Duration getAggregatedTime(final SmackIntegrationTestFramework.TestRunResult testRunResult) {

--- a/src/main/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessor.java
+++ b/src/main/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessor.java
@@ -1,3 +1,18 @@
+/*
+ * Copyright 2024 Guus der Kinderen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package org.igniterealtime.smack.inttest.util;
 
 import org.igniterealtime.smack.inttest.FailedTest;
@@ -22,6 +37,8 @@ import java.io.IOException;
 import java.io.OutputStream;
 import java.lang.reflect.Method;
 import java.net.URI;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -208,11 +225,7 @@ public class JUnitXmlTestRunResultProcessor implements SmackIntegrationTestFrame
                     if (specification != null && !specification.isBlank()) {
                         final Element specificationUrlElement = doc.createElement("property");
                         specificationUrlElement.setAttribute("name", "specification URL");
-                        String link = "https://xmpp.org/extensions/" + specification.toLowerCase() + ".html";
-                        if (specificationSection != null) {
-                            link += "#" + specificationSection; // FIXME this is wrong for XEPs, as they use the title of the section, not its number, as the anchor. Maybe convince someone to add both?
-                        }
-                        specificationUrlElement.setAttribute("value", URI.create(link).toString());
+                        specificationUrlElement.setAttribute("value", generateLink(specification, specificationSection).toString());
                         propertiesElement.appendChild(specificationUrlElement);
                     }
                     testcaseElement.appendChild(propertiesElement);
@@ -236,6 +249,53 @@ public class JUnitXmlTestRunResultProcessor implements SmackIntegrationTestFrame
             }
         } catch (ParserConfigurationException e) {
             throw new RuntimeException(e);
+        }
+    }
+
+    public static URI generateLink(final String specification, final String specificationSection) {
+        if (specification == null || specification.isBlank()) {
+            return null;
+        }
+
+        String link;
+        if (specification.toUpperCase().startsWith("XEP")) {
+            String normalizedSpec = specification.substring(0, 3);
+            switch (specification.charAt(3)) {
+                case ' ':
+                    normalizedSpec += '-' + specification.substring(5);
+                    break;
+                case '-':
+                    normalizedSpec += specification.substring(3);
+                    break;
+                default:
+                    normalizedSpec += '-' + specification.substring(3);
+                    break;
+            }
+            link = "https://xmpp.org/extensions/" + URLEncoder.encode(normalizedSpec.toLowerCase(), StandardCharsets.UTF_8) + ".html";
+            if (specificationSection != null) {
+                link += "#" + URLEncoder.encode(specificationSection, StandardCharsets.UTF_8); // FIXME this is wrong for XEPs, as they use the title of the section, not its number, as the anchor. Maybe convince someone to add both?
+            }
+        } else if (specification.toUpperCase().startsWith("RFC")) {
+            String normalizedSpec = specification.substring(0, 3);
+            switch (specification.charAt(3)) {
+                case '-': // intended fall-through
+                case ' ':
+                    normalizedSpec += specification.substring(4);
+                    break;
+                default:
+                    normalizedSpec += specification.substring(3);
+                    break;
+            }
+            link = "https://www.rfc-editor.org/rfc/" + URLEncoder.encode(normalizedSpec.toLowerCase(), StandardCharsets.UTF_8) + ".html";
+        } else {
+            return null;
+        }
+
+        try {
+            return URI.create(link);
+        } catch (Throwable t) {
+            // Fail to provide a link rather than fail to generate a report.
+            return null;
         }
     }
 

--- a/src/test/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessorTest.java
+++ b/src/test/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessorTest.java
@@ -1,0 +1,211 @@
+/*
+ * Copyright 2024 Guus der Kinderen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.igniterealtime.smack.inttest.util;
+
+import org.junit.jupiter.api.Test;
+
+import java.net.URI;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+/**
+ * Unit tests that verify the implementation of {@link JUnitXmlTestRunResultProcessor}
+ *
+ * @author Guus der Kinderen, guus.der.kinderen@gmail.com
+ */
+public class JUnitXmlTestRunResultProcessorTest
+{
+    @Test
+    public void testNoSpecification() throws Exception
+    {
+        // Setup test fixture.
+        final String specification = null;
+        final String specificationSection = null;
+
+        // Execute system under test.
+        final URI result = JUnitXmlTestRunResultProcessor.generateLink(specification, specificationSection);
+
+        // Verify results.
+        assertNull(result);
+    }
+
+    @Test
+    public void testXepWithSection() throws Exception
+    {
+        // Setup test fixture.
+        final String specification = JUnitXmlTestRunResultProcessor.normalizeSpecification("xep-0030");
+        final String specificationSection = "8";
+
+        // Execute system under test.
+        final URI result = JUnitXmlTestRunResultProcessor.generateLink(specification, specificationSection);
+
+        // Verify results.
+        assertNotNull(result);
+        assertTrue(result.toString().startsWith("https://xmpp.org/extensions/xep-0030.html"));
+    }
+
+    @Test
+    public void testXepWithoutSection() throws Exception
+    {
+        // Setup test fixture.
+        final String specification = JUnitXmlTestRunResultProcessor.normalizeSpecification("xep-0030");
+        final String specificationSection = null;
+
+        // Execute system under test.
+        final URI result = JUnitXmlTestRunResultProcessor.generateLink(specification, specificationSection);
+
+        // Verify results.
+        assertNotNull(result);
+        assertEquals("https://xmpp.org/extensions/xep-0030.html", result.toString());
+    }
+
+    @Test
+    public void testXepWithSpace() throws Exception
+    {
+        // Setup test fixture.
+        final String specification = JUnitXmlTestRunResultProcessor.normalizeSpecification("xep 0030");
+        final String specificationSection = null;
+
+        // Execute system under test.
+        final URI result = JUnitXmlTestRunResultProcessor.generateLink(specification, specificationSection);
+
+        // Verify results.
+        assertNotNull(result);
+        assertEquals("https://xmpp.org/extensions/xep-0030.html", result.toString());
+    }
+
+    @Test
+    public void testXepWithoutSpace() throws Exception
+    {
+        // Setup test fixture.
+        final String specification = JUnitXmlTestRunResultProcessor.normalizeSpecification("xep0030");
+        final String specificationSection = null;
+
+        // Execute system under test.
+        final URI result = JUnitXmlTestRunResultProcessor.generateLink(specification, specificationSection);
+
+        // Verify results.
+        assertEquals("https://xmpp.org/extensions/xep-0030.html", result.toString());
+    }
+
+    @Test
+    public void testXepWithDashSpace() throws Exception
+    {
+        // Setup test fixture.
+        final String specification = JUnitXmlTestRunResultProcessor.normalizeSpecification("xep-0030");
+        final String specificationSection = null;
+
+        // Execute system under test.
+        final URI result = JUnitXmlTestRunResultProcessor.generateLink(specification, specificationSection);
+
+        // Verify results.
+        assertNotNull(result);
+        assertEquals("https://xmpp.org/extensions/xep-0030.html", result.toString());
+    }
+
+    /**
+     * @see <a href="https://github.com/XMPP-Interop-Testing/smack-sint-server-extensions/issues/21">Improper characters being used to construct URL in JUnitXmlTestRunResultProcessor</a>
+     */
+    @Test
+    public void testXepWithFreeFormSection() throws Exception
+    {
+        // Setup test fixture.
+        final String specification = JUnitXmlTestRunResultProcessor.normalizeSpecification("xep-0030");
+        final String specificationSection = "7.1 & 7.2.2";
+
+        // Execute system under test.
+        final URI result = JUnitXmlTestRunResultProcessor.generateLink(specification, specificationSection);
+
+        // Verify results.
+        assertNotNull(result);
+        assertTrue(result.toString().startsWith("https://xmpp.org/extensions/xep-0030.html"));
+    }
+
+    @Test
+    public void testRfcWithSection() throws Exception
+    {
+        // Setup test fixture.
+        final String specification = JUnitXmlTestRunResultProcessor.normalizeSpecification("rfc6121");
+        final String specificationSection = "8";
+
+        // Execute system under test.
+        final URI result = JUnitXmlTestRunResultProcessor.generateLink(specification, specificationSection);
+
+        // Verify results.
+        assertNotNull(result);
+        assertTrue(result.toString().startsWith("https://www.rfc-editor.org/rfc/rfc6121.html"));
+    }
+
+    @Test
+    public void testRfcWithoutSection() throws Exception
+    {
+        // Setup test fixture.
+        final String specification = JUnitXmlTestRunResultProcessor.normalizeSpecification("rfc6121");
+        final String specificationSection = null;
+
+        // Execute system under test.
+        final URI result = JUnitXmlTestRunResultProcessor.generateLink(specification, specificationSection);
+
+        // Verify results.
+        assertNotNull(result);
+        assertEquals("https://www.rfc-editor.org/rfc/rfc6121.html", result.toString());
+    }
+
+    @Test
+    public void testRfcWithSpace() throws Exception
+    {
+        // Setup test fixture.
+        final String specification = JUnitXmlTestRunResultProcessor.normalizeSpecification("rfc 6121");
+        final String specificationSection = null;
+
+        // Execute system under test.
+        final URI result = JUnitXmlTestRunResultProcessor.generateLink(specification, specificationSection);
+
+        // Verify results.
+        assertNotNull(result);
+        assertEquals("https://www.rfc-editor.org/rfc/rfc6121.html", result.toString());
+    }
+
+    @Test
+    public void testRfcWithoutSpace() throws Exception
+    {
+        // Setup test fixture.
+        final String specification = JUnitXmlTestRunResultProcessor.normalizeSpecification("rfc6121");
+        final String specificationSection = null;
+
+        // Execute system under test.
+        final URI result = JUnitXmlTestRunResultProcessor.generateLink(specification, specificationSection);
+
+        // Verify results.
+        assertNotNull(result);
+        assertEquals("https://www.rfc-editor.org/rfc/rfc6121.html", result.toString());
+    }
+
+    @Test
+    public void testRfcWithDashSpace() throws Exception
+    {
+        // Setup test fixture.
+        final String specification = JUnitXmlTestRunResultProcessor.normalizeSpecification("rfc-6121");
+        final String specificationSection = null;
+
+        // Execute system under test.
+        final URI result = JUnitXmlTestRunResultProcessor.generateLink(specification, specificationSection);
+
+        // Verify results.
+        assertNotNull(result);
+        assertEquals("https://www.rfc-editor.org/rfc/rfc6121.html", result.toString());
+    }
+}

